pkg/ui/server_test.go

```diff
@@ -0,0 +1,184 @@
+	expected  bool
+}
+
+type testValidateFormParam struct {
+	detail   FormDetails
+	expected bool
+}
+
+type testIsValidOneOfInputParam struct {
+	selection string
+	options   []string
+	expected  bool
+}
+
+func TestValidateOrigin(t *testing.T) {
+	t.Parallel()
+
+	inputs := []testValidateOriginParam{
+		{
+			origin:    "",
+			allowList: []string{},
+			expected:  false,
+		},
+		{
+			origin:    "",
+			allowList: []string{"foo.com"},
+			expected:  false,
+		},
+		{
+			origin:    "bar.com",
+			allowList: []string{"foo.com"},
+			expected:  false,
+		},
+		{
+			origin:    "go.foo.com",
+			allowList: []string{"bar.com", "baz.com"},
+			expected:  false,
+		},
+		{
+			origin:    "go.foo.com",
+			allowList: []string{"bar.com", "foo.com"},
+			expected:  true,
+		},
+		{
+			origin:    "foo.com",
+			allowList: []string{"bar.com", "foo.com"},
+			expected:  true,
+		},
+		{
+			origin:    "foo.com",
+			allowList: []string{"bar.com", "*"},
+			expected:  true,
+		},
+	}
+
+	for _, value := range inputs {
+		if validateOrigin(value.origin, value.allowList) != value.expected {
+			t.Fatalf("Failed validating %+v", value)
+		}
+	}
+}
+
+func TestValidateForm(t *testing.T) {
+	t.Parallel()
+
+	var inputs []testValidateFormParam
+
+	for i := 0; i < len(categories); i++ {
+		category := categories[i]
+		for j := 0; j < len(ttls); j++ {
+			ttl := ttls[j]
+
+			happyPathCase := testValidateFormParam{
+				detail: FormDetails{
+					Category: category,
+					Reason:   "reason",
+					TTL:      ttl,
+				},
+				expected: true,
+			}
+
+			inputs = append(inputs, happyPathCase)
+		}
+	}
+
+	sadPathCases := []testValidateFormParam{
+		{
+			detail: FormDetails{
+				Category: "",
+				Reason:   "",
+				TTL:      "",
+			},
+			expected: false,
+		},
+		{
+			detail: FormDetails{
+				Category: "",
+				Reason:   "reason",
+				TTL:      ttls[0],
+			},
+			expected: false,
+		},
+		{
+			detail: FormDetails{
+				Category: categories[0],
+				Reason:   "",
+				TTL:      ttls[1],
+			},
+			expected: false,
+		},
+		{
+			detail: FormDetails{
+				Category: categories[0],
+				Reason:   "reason",
+				TTL:      "",
+			},
+			expected: false,
+		},
+	}
+
+	inputs = append(inputs, sadPathCases...)
+
+	for _, value := range inputs {
+		if validateForm(&value.detail) != value.expected {
+			t.Fatalf("Failed validating form %+v", value)
+		}
+	}
+}
+
+func TestIsValidOneOf(t *testing.T) {
+	t.Parallel()
+
+	inputs := []testIsValidOneOfInputParam{
+		{
+			selection: "",
+			options:   []string{},
+			expected:  false,
+		},
+		{
+			selection: "foo",
+			options:   []string{},
+			expected:  false,
+		},
+		{
+			selection: "foo",
+			options:   []string{"bar"},
+			expected:  false,
+		},
+		{
+			selection: "foo",
+			options:   []string{"bar", "foo"},
+			expected:  true,
+		},
+	}
+
+	for _, value := range inputs {
+		if isValidOneOf(value.selection, value.options) != value.expected {
+			t.Fatalf("Failed validating %+v", value)
+		}
+	}
+}
```

Comment:
We need some tests that actually invoke the handler.

Reply:
+1 Not sure if this https://github.com/abcxyz/readability/wiki/Go-Best-Practices#unit-test-real-connections is possible.

Reply:
ack, spent a good amount of time refactoring to follow the patterns in EN. Working on testing the handler next.

Reply:
i added some tests, but i dont love it. After browsing EN, i want to mimic the form handling from there. Maybe as a follow up?